crates/wasm-mutate/src/mutators/peephole/mod.rs

```diff
@@ -0,0 +1,252 @@
+                                innerprev
+                                    .entry(peephole.name())
+                                    .or_insert(Vec::new())
+                                    .push((fidx as usize, idx, peephole));
+                            }
+                        }
+                        innerprev
+                    })
+            });
+
+        // If no mutators, return specific error
+
+        if applicable.len() == 0 {
+            return Err(crate::Error::NotMatchingPeepholes);
+        };
+
+        let mutatoridx = applicable.keys().choose(rnd).unwrap();
+        let positions = &applicable[mutatoridx];
+        let (function_to_mutate, operatoridx, mutator) = positions.choose(rnd).unwrap();
+
+        for fidx in 0..function_count as usize {
+            let mut reader = all_readers[fidx];
+            if fidx == *function_to_mutate {
+                log::debug!("Mutating function idx {:?}", fidx);
+                let function = mutator
+                    .mutate(config, rnd, &mut reader, *operatoridx, &code_section.data)
+                    .unwrap();
+                println!("{:?}", function);
+                codes.function(&function);
+            } else {
+                // Copy exactly the same function to section
+                println!(
+                    "{:?}",
+                    &code_section.data[reader.range().start..reader.range().end]
+                );
+                codes.raw(&code_section.data[reader.range().start..reader.range().end]);
+            }
+        }
+
+        let module = info.replace_section(info.code.unwrap(), &codes);
+        Ok(module)
+    }
+
+    fn can_mutate<'a>(
+        &self,
+        config: &'a crate::WasmMutate,
+        info: &crate::ModuleInfo,
+    ) -> Result<bool> {
+        Ok(info.has_code() && info.function_count > 0)
+    }
+}
+
+use std::fmt::Debug;
+impl Debug for Box<dyn CodeMutator> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Code mutator").finish()
+    }
+}
+pub(crate) trait CodeMutator {
+    fn mutate(
+        &self,
+        config: &WasmMutate,
+        rnd: &mut SmallRng,
+        funcreader: &mut FunctionBody,
+        operator_index: usize,
+        function_data: &[u8],
+    ) -> Result<Function>;
+
+    /// Returns if this mutator can be applied to the opcode at index i
+    fn can_mutate<'a>(
+        &self,
+        config: &'a WasmMutate,
+        operators: &Vec<Operator<'a>>,
+        at: usize,
+    ) -> Result<bool>;
+
+    /// Provides the name of the mutator, mostly used for debugging purposes
+    fn name(&self) -> String {
+        return format!("{:?}", std::any::type_name::<Self>());
+    }
+}
+
+// This macro is meant to be used for testing deep mutators
+// It receives the original wat text variable, the expression returning the mutated function and the expected wat
+// For an example, look at SwapCommutativeOperator
+#[macro_export]
```

Comment:
If this is only for test code, can we add `#[cfg(test)]` on top of it? That way we can't accidentally use it elsewhere, nor will `rustc` have to do anything other than parse it.

Reply:
Done